File: projectforge-wicket/src/main/java/org/projectforge/web/export/ExportZipArchive.java
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2014 Kai Reinhard (k.reinhard@micromata.de)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.web.export;

import java.io.IOException;
import java.io.OutputStream;
import java.util.Collection;
import java.util.LinkedList;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

import org.apache.commons.io.IOUtils;
import org.apache.wicket.util.resource.AbstractResourceStreamWriter;
import org.apache.wicket.util.resource.IResourceStream;
import org.projectforge.business.excel.ExportWorkbook;
import org.projectforge.export.ExportJFreeChart;
import org.projectforge.export.ExportZipFile;

/**
 * For exporting multiple objects by one script you may collect all objects within this zip archive.
 *
 * @author Kai Reinhard (k.reinhard@micromata.de)
 */
public class ExportZipArchive
{
  private static final org.slf4j.Logger log = org.slf4j.LoggerFactory.getLogger(ExportZipArchive.class);

  private final Collection<ExportZipFile> zipFiles = new LinkedList<ExportZipFile>();

  private final String filename;

  public ExportZipArchive()
  {
    this.filename = "archive";
  }

  /**
   * @param filename The filename of the zip archive (without extension), default is "archive".
   */
  public ExportZipArchive(final String filename)
  {
    this.filename = filename;
  }

  public void write(final OutputStream out)
  {
    final ZipOutputStream zipOut = new ZipOutputStream(out);
    try {
      zipOut.putNextEntry(new ZipEntry(filename + "/"));
      for (final ExportZipFile file : zipFiles) {
        final ZipEntry zipEntry = new ZipEntry(filename + "/" + file.getFilename());
        zipOut.putNextEntry(zipEntry);
        if (file.getExportObject() instanceof ExportWorkbook) {
          final ExportWorkbook workbook = (ExportWorkbook) file.getExportObject();
          final byte[] xls = workbook.getAsByteArray();
          if (xls == null || xls.length == 0) {
            log.error("Oups, xls has zero size. Filename: " + filename);
            continue;
          }
          zipOut.write(xls);
        } else if (file.getExportObject() instanceof ExportJFreeChart) {
          final ExportJFreeChart exportJFreeChart = (ExportJFreeChart) file.getExportObject();
          exportJFreeChart.write(zipOut);
        }
        zipOut.closeEntry();
      }
    } catch (final IOException ex) {
      log.error(ex.getMessage(), ex);
      throw new RuntimeException(ex);
    } finally {
      IOUtils.closeQuietly(zipOut);
    }
  }

  public IResourceStream createResourceStreamWriter()
  {
    final IResourceStream iResourceStream = new AbstractResourceStreamWriter()
    {
      private static final long serialVersionUID = 7780552906708508709L;

      @Override
      public String getContentType()
      {
        return "application/zip";
      }

      @Override
      public void write(final OutputStream output)
      {
        ExportZipArchive.this.write(output);
      }
    };
    return iResourceStream;
  }

  public ExportZipArchive add(final String filename, final ExportWorkbook exportWorkbook)
  {
    zipFiles.add(new ExportZipFile(filename, exportWorkbook));
    return this;
  }

  public ExportZipArchive add(final ExportWorkbook exportWorkbook)
  {
    zipFiles.add(new ExportZipFile(exportWorkbook.getFilename(), exportWorkbook));
    return this;
  }

  public ExportZipArchive add(final String filename, final ExportJFreeChart exportJFreeChart)
  {
    zipFiles.add(new ExportZipFile(filename, exportJFreeChart));
    return this;
  }

  /**
   * @return the filename
   */
  public String getFilename()
  {
    return filename;
  }

  public Collection<ExportZipFile> getFiles()
  {
    return zipFiles;
  }
}

File: projectforge-business/src/main/kotlin/org/projectforge/business/scripting/ExportZipArchive.kt
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2026 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.business.scripting

import de.micromata.merlin.excel.ExcelWorkbook
import org.projectforge.business.excel.ExportWorkbook
import org.projectforge.common.ZipMode
import org.projectforge.common.ZipUtils
import org.projectforge.export.ExportJFreeChart
import org.projectforge.export.ExportZipFile
import org.slf4j.LoggerFactory
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.OutputStream
import java.util.*
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream

/**
 * For exporting multiple objects by one script you may collect all objects within this zip archive.
 *
 * @author Kai Reinhard (k.reinhard@micromata.de)
 * @param filename The filename of the zip archive (without extension), default is "archive".
 * @param encryptionPassword Password for encrypting the entire archive (null = no encryption)
 * @param encryptionMode Encryption mode (default: AES-256)
 * @param flatStructure If true, files are added directly to ZIP root without creating a subfolder
 */
class ExportZipArchive(
  val filename: String = "archive",
  var encryptionPassword: String? = null,
  var encryptionMode: ZipMode = ZipMode.ENCRYPTED_AES256,
  val flatStructure: Boolean = false
) {
  private val zipFiles: MutableCollection<ExportZipFile> = LinkedList()

  fun asByteArray(): ByteArray {
    return ByteArrayOutputStream().use { out ->
      write(out)
      out.toByteArray()
    }
  }

  fun write(out: OutputStream) {
    // If encryption is enabled, create unencrypted ZIP first, then encrypt it
    if (encryptionPassword != null) {
      val unencryptedZip = ByteArrayOutputStream()
      writeUnencrypted(unencryptedZip)

      // Encrypt the entire ZIP
      ZipUtils.encryptZipFile(
        "$filename.zip",
        encryptionPassword!!,
        unencryptedZip.toByteArray().inputStream(),
        out,
        encryptionMode
      )
    } else {
      writeUnencrypted(out)
    }
  }

  private fun writeUnencrypted(out: OutputStream) {
    ZipOutputStream(out).use { zipOut ->
      try {
        // Create folder entry only if not using flat structure
        if (!flatStructure) {
          zipOut.putNextEntry(ZipEntry("$filename/"))
        }
        for (file in zipFiles) {
          // Add files directly to root if flatStructure=true, otherwise add inside folder
          val zipEntry = if (flatStructure) {
            ZipEntry(file.filename)
          } else {
            ZipEntry(filename + "/" + file.filename)
          }
          zipOut.putNextEntry(zipEntry)
          file.exportObject?.let { exportObject ->
            if (exportObject is ExportWorkbook) {
              // Older ProjectForge Excel library.
              val xls = exportObject.asByteArray
              if (xls == null || xls.size == 0) {
                log.error("Oups, xls has zero size. Filename: $filename")
                return@let
              }
              zipOut.write(xls)
            } else if (exportObject is ExcelWorkbook) {
              // Newer Merlin Excel library.
              exportObject.use { workbook ->
                val xls: ByteArray = workbook.asByteArrayOutputStream.toByteArray()
                if (xls.isEmpty()) {
                  log.error("Oups, xls has zero size. Filename: $filename")
                  return@let
                }
                zipOut.write(xls)
              }
            } else if (exportObject is ExportJFreeChart) {
              exportObject.write(zipOut)
            } else if (exportObject is String) {
              zipOut.write(exportObject.toByteArray(Charsets.UTF_8))
            } else if (exportObject is ByteArray) {
              zipOut.write(exportObject)
            }
            zipOut.closeEntry()
          }
        }
      } catch (ex: IOException) {
        log.error(ex.message, ex)
        throw RuntimeException(ex)
      }
    }
  }

  fun add(filename: String, exportWorkbook: ExportWorkbook?): ExportZipArchive {
    zipFiles.add(ExportZipFile(filename, exportWorkbook))
    return this
  }

  fun add(exportWorkbook: ExportWorkbook): ExportZipArchive {
    zipFiles.add(ExportZipFile(exportWorkbook.filename, exportWorkbook))
    return this
  }

  fun add(filename: String, excelWorkbook: ExcelWorkbook?): ExportZipArchive {
    excelWorkbook?.pOIWorkbook?.creationHelper?.createFormulaEvaluator()?.evaluateAll()
    zipFiles.add(ExportZipFile(filename, excelWorkbook))
    return this
  }

  fun add(excelWorkbook: ExcelWorkbook): ExportZipArchive {
    zipFiles.add(ExportZipFile(excelWorkbook.filename, excelWorkbook))
    return this
  }

  fun add(filename: String, exportJFreeChart: ExportJFreeChart?): ExportZipArchive {
    zipFiles.add(ExportZipFile(filename, exportJFreeChart))
    return this
  }

  fun add(filename: String, content: String): ExportZipArchive {
    zipFiles.add(ExportZipFile(filename, content))
    return this
  }

  fun add(filename: String, content: ByteArray): ExportZipArchive {
    zipFiles.add(ExportZipFile(filename, content))
    return this
  }

  val files: Collection<ExportZipFile>
    get() = zipFiles

  companion object {
    private val log = LoggerFactory.getLogger(ExportZipArchive::class.java)
  }
}
